kornia/losses/perceptual.py

```diff
@@ -0,0 +1,304 @@
+            C = number of channels
+
+    Return:
+        the corners of the image.
+    """
+    if len(image.shape) != 4:
+        raise ValueError('patch should be of size B, C, H, W')
+    batch_size = image.shape[0]
+    image_width = image.shape[-2]
+    image_height = image.shape[-1]
+    corners = torch.tensor(
+        [[0, 0], [image_width, 0], [image_width, image_height], [0, image_height]],
+        device=image.device,
+        dtype=image.dtype,
+        requires_grad=False,
+    )
+    corners = corners.repeat(batch_size, 1, 1)
+
+    return corners
+
+
+def _four_point_to_homography(corners: torch.Tensor, deltas: torch.Tensor) -> torch.Tensor:
+    """Convert 4-point representation introduced in :cite:`detone2016deep` to homography.
+
+    Args:
+        corners: corners tensor with shape :math:`(B, 4, 2)` where B = batch size
+        deltas: deltas tensor with shape :math:`(B, 4, 2)` where B = batch size
+
+    Return:
+        the converted homography.
+    """
+
+    if not isinstance(corners, torch.Tensor):
+        raise TypeError(f"corners type is not a torch.Tensor. Got {type(corners)}")
+
+    if not isinstance(deltas, torch.Tensor):
+        raise TypeError(f"deltas type is not a torch.Tensor. Got {type(deltas)}")
+
+    if not len(corners.shape) == 3 or not corners.shape[1] == 4 or not corners.shape[2] == 2:
+        raise ValueError(f"Invalid input shape of corners, we expect Bx4x2. Got: {corners.shape}")
+
+    if not len(deltas.shape) == 3 or not deltas.shape[1] == 4 or not deltas.shape[2] == 2:
+        raise ValueError(f"Invalid input shape of deltas, we expect Bx4x2. Got: {deltas.shape}")
+
+    if not corners.size(0) == deltas.size(0):
+        raise ValueError(f'Expected corners batch_size ({corners.size(0)}) to match deltas batch '
+                         f'size ({deltas.size(0)}).')
+
+    corners_hat = corners + deltas
+    homography_inv = get_perspective_transform(corners, corners_hat)
+    homography = torch.inverse(homography_inv)
+    return homography
+
+
+def _warp(image: torch.Tensor, delta_hat: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+    """Convert 4-point representation introduced in :cite:`detone2016deep` to homography.
+
+    Args:
+        image: image tensor with shape :math:`(B, C, H, W)` where B = batch size,
+            C = number of channels
+        deltas: deltas tensor with shape :math:`(B, 4, 2)` where B = batch size
+
+    Return:
+        the warped images.
+    """
+    corners = _image_shape_to_corners(image=image)
+    homography = _four_point_to_homography(corners=corners, deltas=delta_hat)
+    image_warped = warp_perspective(image, homography, (image.shape[-2], image.shape[-1]))
+    return image_warped, torch.inverse(homography)
+
+
+def bihome_loss(
+    patch_1: torch.Tensor,
+    patch_2: torch.Tensor,
+    delta_hat_12: torch.Tensor,
+    delta_hat_21: torch.Tensor,
+    triplet_mu: float,
+    loss_network: nn.Module,
```

Comment:
```suggestion
    features_fcn: Optional[Callable] = None,
```
not very intuitive the naming since i'd expected a network that returns my a loss already which i believe it's not the case

Later in the code:

```python
if features_fcn is None:
  features_fcn = nn.Identity()
```

Reply:
I agree that `loss_network` is not the most informative name, but I think it's just a convention used by a very known paper about perceptual loss: https://arxiv.org/abs/1603.08155.

Another question is if we really want to let the user use Identity as loss_network. Then it's not a perceptual loss anymore, but normal photometric loss. Idk, I think photometric loss is kind of a special case of perceptual loss, but I am not sure if this won't introduce some silent "bugs" for possible users.